### DISCUSSION TITLE
Trigger dag immediately with rest API

### DISCUSSION THREAD

**Discussion by fzhangcsc** [apache/airflow#31426] [Q&A]
How could I config a dag  to trigger immediately and only once by rest API

I tried to set 

start_date = datetime(2023, 5, 18),
schedule_interval=None,

but still only see the dag is scheduled at airflow UI  when triggered by rest API, but this dag is not executed.

I still need manually trigger this dag to execute tasks for this dag.

Thanks
Fan

**Comment by potiuk** [comment]
Did you try ` schedule="@once"` ?

**Comment by fzhangcsc** [comment]
not working
dag = DAG(
    'save_config_dag',
    description='An event-based DAG that is triggered by an HTTP POST request to save request meta data',
    start_date = datetime(2023, 5, 18),
    schedule='@once',
    concurrency=1,
    max_active_runs=1,
    tags=['experiment']
)

the tasks under this dag are scheduled but not executed

@potiuk 

response from rest API trigger call


{
  "conf": {},
  "dag_id": "save_config_dag",
  "dag_run_id": "manual__2023-05-20T00:54:56.302821+00:00",
  "data_interval_end": "2023-05-20T00:54:56.302821+00:00",
  "data_interval_start": "2023-05-20T00:54:56.302821+00:00",
  "end_date": null,
  "execution_date": "2023-05-20T00:54:56.302821+00:00",
  "external_trigger": true,
  "last_scheduling_decision": null,
  "logical_date": "2023-05-20T00:54:56.302821+00:00",
  "note": null,
  "run_type": "manual",
  "start_date": null,
  "state": "queued"
}

Thanks

**Comment by fzhangcsc** [comment]
Please help to resolve unexpected scheduling issue.

We want all dag tasks are executed immediately when trigged by rest API instead of queued

**Reply by potiuk** [reply]
You can `want it` a lot but it's not how it works. The only entity that can schedule the DAG is scheduler. It needs to be running and it needs to schedule the task according to it's configuration, limits, other DAGs etc. When you trigger  a DAG via rest API you do exactly what you see-> you put the task in the queued state, and scheduler should pick it up and run it when all the conditions for running the dags are fullfilled. 

If your DAGs/tasks are queued but not running quickly you should make sure your scheduler is running, you should look [why your tasks are not being scheduled](https://airflow.apache.org/docs/apache-airflow/stable/faq.html#why-is-task-not-getting-scheduled) and possibly [fine tune your scheduler performance](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/scheduler.html#fine-tuning-your-scheduler-performance). Also maybe you have some conditions in your DAG/tasks that prevent parallel tasks from running - hard to say. 

Without you looking at why your task is not being scheduled, there is not much anyone can help you with.

For now I am not sure whats' not working for you - your description is a little vague. You mention that triggered dags do not work, but manual triggering works, which is rather strange because triggering via REST API does exactly the same (setting task to queued) as manually triggering it in UI. Also explaining your Airflow version and deployment would be useful for anyone who might want to help you.

If you have more concrete logs and screenshots and describe what exactly problem you have maybe someone will be able to help you.

**Comment by zoid-w** [comment]
Hi, I have designed some pipelines that do this without issue.

Most likely your DAG is indeed in the paused state (assuming there is no issue on your scheduler) as @potiuk referenced.
Either activate your DAG manually or use the patch endpoint to unpause it before the POST request triggering a new dagrun is done: 

`
  req_data_unpause = {
      "is_paused": False
  }
  
  unpause_resp = requests.patch(
      url="{{YOUR AIRFLOW}}/api/v1/dags/{{dag_id}}",
      json=req_data_unpause,
      auth=(username, password)
  )
`

**Comment by fzhangcsc** [comment]
Tried, not working 


Sent from Yahoo Mail for iPhone


On Friday, May 19, 2023, 11:43 AM, Jarek Potiuk ***@***.***> wrote:




Did you try  ***@***.***" ?

**Comment by fzhangcsc** [comment]
Great.
I could not find document to use requests.patch in airflow API document.
Could you show me the link of doc? or give me a full example to trigger restful API with patch 
I am using the following curl command to trigger rest API
  airflowjsonObj = {'conf':{**json.loads(jsonObj)}}
  data = json.dumps(airflowjsonObj)        curlCommand = "curl -X POST 'http://localhost:8080/api/v1/dags/event_based_dag/dagRuns' -H 'Cache-Control: no-cache' -H 'content-type: application/json' --user 'admin2:admin2' --data '"  + data +"'"

 jsonObj is the data to send to trigged dag.
Everything works as expected except that dag is paused.

Should I change POST to Patch?
Also how to add req_data_unpause = {
"is_paused": False
} into curl command?

Thanks
Fan    On Wednesday, May 24, 2023 at 01:02:16 AM PDT, zoid-w ***@***.***> wrote:  
 
 


Hi, I have designed some pipelines that do this without issue.

Most likely your DAG is indeed in the paused state (assuming there is no issue on your scheduler) as @potiuk referenced.
Either activate your DAG manually or use the patch endpoint to unpause it before the POST request triggering a new dagrun is done:

`req_data_unpause = {
"is_paused": False
}

unpause_resp = requests.patch(
url="{{YOUR AIRFLOW}}/api/v1/dags/{{dag_id}}",
json=req_data_unpause,
auth=(username, password)
)
`

**Comment by fzhangcsc** [comment]
I set     ***@***.***' when creating dag
after applying {is_paused: false} in rest API call

I got 
"timetable_description": "Never, external triggers only" from rest API response 

in airflow UI

task unser dag is created 

when when I mouse over the task
I got 
status:queuedtype:manual 
Still need do manual trigger to run this task instead of running the task immediately when associated dag is triggered.

Please let me know if I had wrong setting 

ThanksFan



    On Friday, June 2, 2023 at 02:36:43 PM PDT, FAN ZHANG ***@***.***> wrote:  
 
  Great.
I could not find document to use requests.patch in airflow API document.
Could you show me the link of doc? or give me a full example to trigger restful API with patch 
I am using the following curl command to trigger rest API
  airflowjsonObj = {'conf':{**json.loads(jsonObj)}}
  data = json.dumps(airflowjsonObj)        curlCommand = "curl -X POST 'http://localhost:8080/api/v1/dags/event_based_dag/dagRuns' -H 'Cache-Control: no-cache' -H 'content-type: application/json' --user 'admin2:admin2' --data '"  + data +"'"

 jsonObj is the data to send to trigged dag.
Everything works as expected except that dag is paused.

Should I change POST to Patch?
Also how to add req_data_unpause = {
"is_paused": False
} into curl command?

Thanks
Fan    On Wednesday, May 24, 2023 at 01:02:16 AM PDT, zoid-w ***@***.***> wrote:  
 
 


Hi, I have designed some pipelines that do this without issue.

Most likely your DAG is indeed in the paused state (assuming there is no issue on your scheduler) as @potiuk referenced.
Either activate your DAG manually or use the patch endpoint to unpause it before the POST request triggering a new dagrun is done:

`req_data_unpause = {
"is_paused": False
}

unpause_resp = requests.patch(
url="{{YOUR AIRFLOW}}/api/v1/dags/{{dag_id}}",
json=req_data_unpause,
auth=(username, password)
)
`

**Comment by pankajastro** [comment]
As mentioned above most likely your DAG is in the paused state you can quickly confirm that from UI. Your DAG should not be in a pause state. 
<img width="266" alt="Screenshot 2023-06-04 at 12 46 10 AM" src="https://github.com/apache/airflow/assets/98807258/c78e68f0-1adf-44f1-a34f-7a6a9bc76087">


If it is in the paused state you can unpause it either from UI or with
```
curl -X POST http://localhost:28080/api/v1/dags/$DAG_ID/dagRuns --header "Content-Type: application/json" --header "Authorization: Basic $BASIC_AUTH" -d {}
```
the trigger it with

```
curl -X PATCH http://localhost:28080/api/v1/dags/$DAG_ID--header "Content-Type: application/json" --header "Authorization: Basic $BASIC_AUTH" -d '{"is_paused": false}'
```
You will have to update BASIC_AUTH and DAG_ID in the above curl request.

equivalent python API you can find here to unpaused DAG
https://github.com/apache/airflow-client-python/blob/main/airflow_client/client/api/dag_api.py#L1301
 
API docs 
- [Patch DAG](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/patch_dags)
- [Trigger DAG](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/post_dag_run)

**Comment by fzhangcsc** [comment]
jsonObj = json.dumps(request.json) airflowjsonObj = {'conf':{**json.loads(jsonObj)}} data = json.dumps(airflowjsonObj)
curl -X POST http://localhost:28080/api/v1/dags/$DAG_ID/dagRuns --header "Content-Type: application/json" --header "Authorization: Basic $BASIC_AUTH"  --data '"  + data +"'"

the trigger it with
curl -X PATCH http://localhost:28080/api/v1/dags/$DAG_ID--header "Content-Type: application/json" --header "Authorization: Basic $BASIC_AUTH" -d '{"is_paused": false}'
Helps to unpause dag
However, with above implementation, 
triggered dag does not honor the data sent from curl -X POST http://localhost:28080/api/v1/dags/$DAG_ID/dagRuns --header "Content-Type: application/json" --header "Authorization: Basic $BASIC_AUTH"  --data '"  + data +"'"
instead the trigged dag always uses saved data when executing dag.
If I skip curl -X PATCH http://localhost:28080/api/v1/dags/$DAG_ID--header "Content-Type: application/json" --header "Authorization: Basic $BASIC_AUTH" -d '{"is_paused": false}'
I need to manually trigger dag for $DAG_ID. But at this moment, executed dag honors conf setting data.
Please help to resolve side effect of unpausing dag that does honor updated data attached in trigger command 
via curl -X POST http://localhost:28080/api/v1/dags/$DAG_ID/dagRuns --header "Content-Type: application/json" --header "Authorization: Basic $BASIC_AUTH"  --data '"  + data +"'"
ThanksFan
 

    On Saturday, June 3, 2023 at 12:23:42 PM PDT, Pankaj Singh ***@***.***> wrote:  
 
 


As mentioned above most likely your DAG is in the paused state you can quickly confirm that from UI. Your DAG should not be in a pause state.


If it is in the paused state you can unpause it either from UI or with
curl -X POST http://localhost:28080/api/v1/dags/$DAG_ID/dagRuns --header "Content-Type: application/json" --header "Authorization: Basic $BASIC_AUTH" -d {}

the trigger it with
curl -X PATCH http://localhost:28080/api/v1/dags/$DAG_ID--header "Content-Type: application/json" --header "Authorization: Basic $BASIC_AUTH" -d '{"is_paused": false}'

You will have to update BASIC_AUTH and DAG_ID in the above curl request.

equivalent python API you can find here to unpaused DAG
https://github.com/apache/airflow-client-python/blob/main/airflow_client/client/api/dag_api.py#L1301
